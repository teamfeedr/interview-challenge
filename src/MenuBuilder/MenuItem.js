import React from 'react';
import DieteriesList from './DietariesList';

const MenuItem = ({ id, name, dietaries, remove = () => {} }) => (
  <li className="item">
    <h2>{name}</h2>
    <DieteriesList dietaries={dietaries} />
    {
      <button className="remove-item" onClick={() => remove()}>
        x
      </button>
    }
  </li>
);

export default MenuItem;
