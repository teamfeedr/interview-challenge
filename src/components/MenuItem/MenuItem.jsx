import React from 'react';
import PropTypes from 'prop-types';

import DietaryIcon from '../DietaryIcon/DietaryIcon';

const getDietaryIconElements = ({ dietaries }) =>
    dietaries.map((diet) => (<DietaryIcon diet={diet} key={diet} />));

const MenuItem = ({ handleOnSelect, handleOnDeselect, dietaries, name, id }) => {
    const dietaryIcons = getDietaryIconElements({ dietaries });
    const showRemoveIcon = handleOnDeselect;

    return (
        <li key={id} className="item" onClick={handleOnSelect}>
            <h2>{name}</h2>
            <p>
                {dietaryIcons}
            </p>
            {showRemoveIcon && (
                <button className="remove-item" onClick={handleOnDeselect}>x</button>
            )}
        </li>
    )
}

MenuItem.propTypes = {
    dietaries: PropTypes.arrayOf(PropTypes.string).isRequired,
    name: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    handleOnDeselect: PropTypes.func,
    handleOnSelect: PropTypes.func
}

MenuItem.defaultProps = {
    handleOnDeselect: undefined,
    handleOnSelect: undefined
}

export default MenuItem;