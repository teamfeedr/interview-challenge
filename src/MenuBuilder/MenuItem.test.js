import React from 'react';
import MenuItem from './MenuItem';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders empty', () => {
  const div = shallow(<MenuItem />);
  expect(div).toMatchSnapshot();
});

it('renders food details', () => {
  const div = shallow(
    <MenuItem id={1001} name="Yummy food" dietaries={['v', 'n!']} />
  );
  expect(div).toMatchSnapshot();
});

it('calls remove on click of the X', () => {
  const spy = jest.fn();

  const div = shallow(
    <MenuItem
      id={1001}
      name="Yummy food"
      dietaries={['v', 'n!']}
      remove={spy}
    />
  );
  div.find('.remove-item').simulate('click', {
    target: {
      parentElement: { id: 5 },
      id: 6
    }
  });

  expect(spy).toHaveBeenCalled();
});
