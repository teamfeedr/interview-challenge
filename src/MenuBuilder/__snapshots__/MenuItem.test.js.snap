// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuItem />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2 />,
        <DieteriesList
          dietaries={undefined}
/>,
        <button
          className="remove-item"
          onClick={[Function]}
>
          x
</button>,
      ],
      "className": "item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "dietaries": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "x",
          "className": "remove-item",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "x",
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2 />,
          <DieteriesList
            dietaries={undefined}
/>,
          <button
            className="remove-item"
            onClick={[Function]}
>
            x
</button>,
        ],
        "className": "item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "dietaries": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "x",
            "className": "remove-item",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "x",
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renders food details 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuItem
    dietaries={
        Array [
            "v",
            "n!",
          ]
    }
    id={1001}
    name="Yummy food"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Yummy food
</h2>,
        <DieteriesList
          dietaries={
                    Array [
                              "v",
                              "n!",
                            ]
          }
/>,
        <button
          className="remove-item"
          onClick={[Function]}
>
          x
</button>,
      ],
      "className": "item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Yummy food",
        },
        "ref": null,
        "rendered": "Yummy food",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "dietaries": Array [
            "v",
            "n!",
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "x",
          "className": "remove-item",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "x",
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Yummy food
</h2>,
          <DieteriesList
            dietaries={
                        Array [
                                    "v",
                                    "n!",
                                  ]
            }
/>,
          <button
            className="remove-item"
            onClick={[Function]}
>
            x
</button>,
        ],
        "className": "item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Yummy food",
          },
          "ref": null,
          "rendered": "Yummy food",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "dietaries": Array [
              "v",
              "n!",
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "x",
            "className": "remove-item",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "x",
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
